cluster/cluster.go

```diff
@@ -563,3 +571,101 @@ func MustGetSegmentConfiguration(connection *dbconn.DBConn, getMirrors ...bool)
 	gplog.FatalOnError(err)
 	return segConfigs
 }
+
+// GetSegmentConfigurationFromFile function to read segment configuration from dumpconfiguration file.
+// file should have the absolute path else if the file is empty it will generate default path coordinator_data_dir/gpsegconfig_dump
+// file should have the valid configuration format
+// dbid content role prole mode status port hostname address datadir
+func GetSegmentConfigurationFromFile(file string) ([]SegConfig, error) {
+	var conFilePath string
+	if file == "" {
+		filename := "gpsegconfig_dump" // filename which dump from fts process to master datadir
+		coordinator_datadir := os.Getenv("COORDINATOR_DATA_DIRECTORY")
+		if coordinator_datadir == "" {
+			coordinator_datadir = os.Getenv("MASTER_DATA_DIRECTORY")
+		}
+		if coordinator_datadir == "" {
+			errStr := "COORDINATOR_DATA_DIRECTORY is not set, not able to generate gpsegconfig_dump file path"
+			gplog.Error(errStr)
+			return nil, fmt.Errorf(errStr)
+		}
+		conFilePath = filepath.Join(coordinator_datadir, filename)
+	} else {
+		conFilePath = file
+	}
+
+	fd, err := os.Open(conFilePath)
+	if err != nil {
+		errStr := fmt.Sprintf("failed opening file %s : %s", conFilePath, err.Error())
+		gplog.Error(errStr)
+		return nil, fmt.Errorf(errStr)
+	}
+	defer fd.Close()
+
+	results := make([]SegConfig, 0)
+	scanner := bufio.NewScanner(fd)
+
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		parts := len(fields)
+		var datadir string
+
+		/* older version of gpsegconfig_dump file has 9 parts as it doesn't have datadir
+		1 -1 p p n u 7000 shrakeshSMD6M.vmware.com shrakeshSMD6M.vmware.com */
+		/* newer version of gpsegconfig_dump file has 10 parts as it does have datadir
+		1 -1 p p n u 7000 shrakeshSMD6M.vmware.com shrakeshSMD6M.vmware.com /data/qddir/demoDataDir-1 */
+		if parts != 9 && parts != 10 {
+			errStr := fmt.Sprintf("Unexpected number of fields (%d) in line: %s", parts, scanner.Text())
+			gplog.Error(errStr)
+			return nil, fmt.Errorf(errStr)
+		}
+
+		dbID, err := strconv.Atoi(fields[0])
```

Comment:
You can create a common function for the conversion
```
func convertField(fieldName string, fieldValue string) (int, error) {
	val, err := strconv.Atoi(fieldValue)
	if err != nil {
		errStr := fmt.Sprintf("Failed to convert %s with value %s to an int. Error: %s", fieldName, fieldValue, err.Error())
		gplog.Error(errStr)
		return -math.MaxInt, fmt.Errorf(errStr)
	}
	return val, nil
}
```

Reply:
> You can create a common function for the conversion
> 
> ```
> func convertField(fieldName string, fieldValue string) (int, error) {
> 	val, err := strconv.Atoi(fieldValue)
> 	if err != nil {
> 		errStr := fmt.Sprintf("Failed to convert %s with value %s to an int. Error: %s", fieldName, fieldValue, err.Error())
> 		gplog.Error(errStr)
> 		return -math.MaxInt, fmt.Errorf(errStr)
> 	}
> 	return val, nil
> }
> ```

@kaknikhil I am confused about where should i keep the convertField function. 

1. As an anonymous function 
` convertField := func (fieldName string, fieldValue string) (int, error){`
 this way the scope of the function would be limited and specific to GetSegmentConfigurationFromFile()
 
 2. As a utility function:
    one option is that I need to write the function and test cases for AtoI, ItoA, and similar conversions to complete the utility which will increase the scope of the pr, and also in the common-go-libs currently, there is no specific package for conversion utility and I feel it doesn't fit as well in the gp libs as the caller can take care of conversion and related error message. 
    Another option is to keep it as the unexposed function in the cluster package. 
    
I feel keeping the function anonymous would be good to go, let me know your opinion.

Reply:
Let me think about this a bit but think option 1 should be fine